sdk/go/auto/cmd.go

```diff
@@ -17,17 +17,203 @@ package auto
 import (
 	"bytes"
 	"context"
+	"fmt"
 	"io"
+	"net/http"
 	"os"
 	"os/exec"
+	"path"
+	"runtime"
+	"strings"
 
+	"github.com/blang/semver"
+	"github.com/pulumi/pulumi/sdk/v3"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/env"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/slice"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/httputil"
 )
 
 const unknownErrorCode = -2
 
-func runPulumiCommandSync(
-	ctx context.Context,
+// PulumiCommand manages the Pulumi CLI and runs operations.
+type PulumiCommand interface {
+	Run(ctx context.Context,
+		workdir string,
+		stdin io.Reader,
+		additionalOutput []io.Writer,
+		additionalErrorOutput []io.Writer,
+		additionalEnv []string,
+		args ...string) (string, string, int, error)
+	Version() semver.Version
+}
+
+type pulumiCommand struct {
+	version semver.Version
+	command string
+}
+
+// NewPulumiCommand creates a new PulumiCommand instance that uses a local CLI
+// installation. PulumiCommandOptions can be used to configure things like the
+// the installation root or to validate that the version meets a minimum
+// requirement. By default the pulumi binary found in $PATH is used.
+func NewPulumiCommand(opts ...PulumiCommandOption) (PulumiCommand, error) {
+	pOpts := &pulumiCommandOptions{}
+	for _, opt := range opts {
+		opt.applyPulumiCommandOption(pOpts)
+	}
+
+	command := "pulumi"
+	if pOpts.root != "" {
+		command = path.Join(pOpts.root, "bin", "pulumi")
+	}
+
+	cmd := exec.Command(command, "version")
+	out, err := cmd.Output()
+	if err != nil {
+		return pulumiCommand{}, fmt.Errorf("failed to run `pulumi version`: %w", err)
+	}
+	currentVersion := strings.TrimSpace(string(out))
+	min := minimumVersion
+	if pOpts.version.GT(min) {
+		min = pOpts.version
+	}
+	version, err := parseAndValidatePulumiVersion(min, currentVersion, pOpts.skipVersionCheck)
+	if err != nil {
+		return pulumiCommand{}, err
+	}
+
+	return pulumiCommand{
+		version: version,
+		command: command,
+	}, nil
+}
+
+// InstallPulumiCommand downloads and installs the Pulumi CLI. By default the
+// CLI version matching the current SDK release is installed in
+// $HOME/.pulumi/versions/$VERSION. Use the option `PulumiCommandRoot` to
+// specify a different directory, and `PulumiCommandVersion` to install a
+// custom version.
+func InstallPulumiCommand(ctx context.Context, opts ...PulumiCommandOption) (PulumiCommand,
+	error,
+) {
```

Comment:
Feel free to ignore this, but the code formatting feels a bit odd here.  Maybe 
```suggestion
func InstallPulumiCommand(ctx context.Context, opts ...PulumiCommandOption) (
	PulumiCommand,
	error,
) {
```

would be slightly better, or keeping it all on one line (if the line doesn't end up too long for the linter)